app/services/katello/content_unit_indexer.rb

```diff
@@ -142,14 +142,42 @@ def fetch_only_ids
     def sync_repository_associations(assocication_tracker, additive: false)
       unless additive
         ActiveRecord::Base.connection.uncached do
-          @model_class.repository_association_class.where(repository_id: @repository.id).where.
-            not(@model_class.unit_id_field => assocication_tracker.unit_ids).delete_all
+          repo_associations_to_destroy = @model_class.repository_association_class.where(repository_id: @repository.id).where.
+            not(@model_class.unit_id_field => assocication_tracker.unit_ids)
+          clean_filter_rules(repo_associations_to_destroy) if repo_associations_to_destroy.present? && [::Katello::ModuleStream, ::Katello::Erratum, ::Katello::PackageGroup].include?(@model_class)
+          repo_associations_to_destroy.destroy_all
         end
       end
       return if assocication_tracker.db_values.empty?
       @model_class.repository_association_class.upsert_all(assocication_tracker.db_values, :unique_by => association_class_uniqiness_attributes)
     end
 
+    def clean_filter_rules(repo_associations_to_destroy)
+      affected_content_view_ids = @repository.content_views.non_default.pluck(:id)
+      case @model_class.to_s
+      when 'Katello::ModuleStream'
+        module_stream_ids = repo_associations_to_destroy.pluck(:module_stream_id)
+        filter_rules = ::Katello::ContentViewModuleStreamFilterRule.
+          joins('INNER JOIN katello_content_view_filters ON katello_content_view_module_stream_filter_rules.content_view_filter_id = katello_content_view_filters.id').
+          where("katello_content_view_filters.content_view_id IN (#{affected_content_view_ids.join(',')})").where(module_stream_id: module_stream_ids)
+        filter_rules.delete_all
+      when 'Katello::Erratum'
+        errata_ids = ::Katello::Erratum.where(id: repo_associations_to_destroy.select(:erratum_id)).pluck(:errata_id)
+        filter_rules = ::Katello::ContentViewErratumFilterRule.
+          joins('INNER JOIN katello_content_view_filters ON katello_content_view_erratum_filter_rules.content_view_filter_id = katello_content_view_filters.id').
+          where("katello_content_view_filters.content_view_id IN (#{affected_content_view_ids.join(',')})").where(errata_id: errata_ids)
+        filter_rules.delete_all
+      when 'Katello::PackageGroup'
+        package_group_uuids = ::Katello::PackageGroup.where(id: repo_associations_to_destroy.select(:package_group_id)).pluck(:pulp_id)
+        filter_rules = ::Katello::ContentViewPackageGroupFilterRule.
+          joins('INNER JOIN katello_content_view_filters ON katello_content_view_package_group_filter_rules.content_view_filter_id = katello_content_view_filters.id').
+          where("katello_content_view_filters.content_view_id IN (#{affected_content_view_ids.join(',')})").where(uuid: package_group_uuids)
+        filter_rules.delete_all
+      else
+        return false
```

Comment:
We have a helper method "rules" for content view filters that is a generic method to return whatever rule class exists on the filter of any type. ` Katello::ContentViewFilter.find(37).rules` 
Would using that simplify the code here and have everything handled with ActiveRecord? I am not against raw SQL but it hurts my brain more than ActiveRecord.

Reply:
Maybe, I'll check if it's applicable

Reply:
I'm not sure that's useful here since I need to collect rules for an arbitrary number of content view filters.

Reply:
Also I was thinking a task that does this on any given CV without access to indexing data might be helpful. If a module stream exists as a rule in a CV but the library repo has no association to the module stream, we delete it. Shouldn't be a long running task for a single repo CVs especially if we make it async in the finalize. Added benefit would be we can run it directly on CVs if needed to clean up filters rather than rely on indexing action on repo.

Reply:
If we make a rake task I could include it as an upgrade task I suppose. We'll just need to be very sure that it works well since we're deleting things.